### PR TITLE
Reparameterize Scrollable2

### PR DESCRIPTION
**Pull request by abarth**
This patch makes a number of changes to how you can configure a
Scrollable2:

 - The ScrollPhysics is now responsible for creating the ScrollPosition.
   You can override the ScrollPhysics by supplying a `physics` argument
   to `Scrollable`, and the new physics you supply will be applied to
   the default physics inherited from the ScrollBehavior.

 - This patch removes the ScrollPosition/AbsoluteScrollPosition split as
   all clients were operating in pixels anyway and the split made the
   code very difficult to follow.

 - ScrollPosition no longer depends directly on Scrollable2State.
   Instead, it depends on an abstract interface that Scrollable2State
   implements. This change has two benefits:

    a) It removes the circular dependency between ScrollPosition and
       Scrollable2State, which lets us split the code for these classes
       (and several other classes that got wrapped up in that cycle) into
       separate libraries for easier maintenance.

    b) ScrollPosition is no longer bound to Scrollable2, which means you
       could use the behavior machinery to drive other sorts of widgets.
       For example, we could use it to drive Scrollabe1 if we wanted.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie , as discussed.

**Comment by abarth** [comment]
Fixes #7833

File: packages/flutter/lib/src/widgets/scroll_configuration.dart [hunk 1]
```diff
@@ -2,12 +2,86 @@
...
+  /// The platform whose scroll physics should be implemented.
+  ///
+  /// Defaults to the current platform.
+  TargetPlatform getPlatform(BuildContext context) => defaultTargetPlatform;
+
+  /// The color to use for the glow effect when [platform] indicates a platform
+  /// that uses a [GlowingOverscrollIndicator].
+  ///
+  /// Defaults to white.
+  Color getGlowColor(BuildContext context) => const Color(0xFFFFFFFF);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart]: we should move this to an inherited widget that is wrapped around the glow by the material layer (the glow itself coming from this layer)

File: packages/flutter/lib/src/widgets/scroll_physics.dart [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        end = position.minScrollExtent;
+      assert(end != null);
+      return new ScrollSpringSimulation(spring, position.pixels, position.maxScrollExtent, math.min(0.0, velocity))
+        ..tolerance = tolerance;
+    }
+    if (!position.atEdge && velocity.abs() >= tolerance.velocity) {
+      return new ClampingScrollSimulation(
+        position: position.pixels,
+        velocity: velocity,
+      )..tolerance = tolerance;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_physics.dart]: why can't you pass the tolerance to the constructor?

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 3]
```diff
@@ -15,202 +15,244 @@ import 'package:flutter/scheduler.dart';
...
-  }
-
-  @override
-  double applyPhysicsToUserOffset(AbsoluteScrollPosition position, double offset) {
+  SpringDescription get spring {
     if (parent == null)
-      return super.applyPhysicsToUserOffset(position, offset);
-    return parent.applyPhysicsToUserOffset(position, offset);
+      return _kDefaultSpring;
+    return parent.spring;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: i'm totally fine with this the way you wrote it, but i'm sure you would tell me to use `parent?.spring ?? _kDefaultSpring` :-)

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 4]
```diff
@@ -15,202 +15,244 @@ import 'package:flutter/scheduler.dart';
...
-class AbsoluteScrollPosition extends ScrollPosition {
-  AbsoluteScrollPosition(
-    Scrollable2State state,
-    Tolerance scrollTolerances,
-    ScrollPosition oldPosition,
-    this.physics,
-  ) : super(state, scrollTolerances, oldPosition);
+class ScrollPosition extends ViewportOffset {
+  ScrollPosition(this.physics, this.state, ScrollPosition oldPosition) {
+    assert(state.vsync != null);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: assert physics != null?

File: packages/flutter/lib/src/widgets/scroll_position.dart [hunk 5]
```diff
@@ -287,259 +334,229 @@ class AbsoluteScrollPosition extends ScrollPosition {
...
+    assert(other.state == state);
+    assert(other.activity != null);
+
+    _pixels = other._pixels;
+    _viewportDimension = other.viewportDimension;
+    _minScrollExtent = other.minScrollExtent;
+    _maxScrollExtent = other.maxScrollExtent;
+
+    final bool oldIgnorePointer = shouldIgnorePointer;
+    _userScrollDirection = other._userScrollDirection;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: we should probably copy this one over before reading shouldIgnorePointer

**Comment by Hixie** [comment]
LGTM.

I think we'll eventually split ScrollPosition into two again, so that we can create delegate ScrollPositions. But the split will be different than what we had before so this is fine.

File: packages/flutter/lib/src/widgets/scroll_physics.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_physics.dart]: Done.

File: packages/flutter/lib/src/widgets/scroll_position.dart (hunk 3 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_position.dart]: Done.

File: packages/flutter/lib/src/widgets/scroll_configuration.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/scroll_configuration.dart]: I've left this for a future patch.

**Pull request merged**